votor/src/consensus_pool_service.rs

```diff
@@ -412,3 +412,269 @@ impl ConsensusPoolService {
+        let exit = Arc::new(AtomicBool::new(false));
+        let leader_schedule_cache = Arc::new(LeaderScheduleCache::new_from_bank(&root_bank.load()));
+        let ctx = ConsensusPoolContext {
+            exit: exit.clone(),
+            start: Arc::new((Mutex::new(true), Condvar::new())),
+            cluster_info: Arc::new(cluster_info),
+            my_vote_pubkey: Pubkey::new_unique(),
+            blockstore: Arc::new(blockstore),
+            root_bank: root_bank.clone(),
+            leader_schedule_cache: leader_schedule_cache.clone(),
+            consensus_message_receiver,
+            bls_sender,
+            event_sender,
+            commitment_sender,
+        };
+        ConsensusPoolServiceTestComponents {
+            consensus_pool_service: ConsensusPoolService::new(ctx),
+            consensus_message_sender,
+            bls_receiver,
+            event_receiver,
+            commitment_receiver,
+            validator_keypairs,
+            exit,
+            leader_schedule_cache,
+            root_bank,
+        }
+    }
+
+    fn wait_for_event<T>(receiver: &Receiver<T>, condition: impl Fn(&T) -> bool) {
+        let start = Instant::now();
+        let mut event_received = false;
+        while start.elapsed() < Duration::from_secs(5) {
+            if let Ok(event) = receiver.recv_timeout(Duration::from_millis(500)) {
+                if condition(&event) {
+                    event_received = true;
+                    break;
+                }
+            }
+        }
+        assert!(event_received);
+    }
+
+    #[test]
+    fn test_receive_and_send_consensus_message() {
+        solana_logger::setup();
+        let setup_result = setup();
+        let consensus_message_sender = setup_result.consensus_message_sender;
+        let consensus_pool_service = setup_result.consensus_pool_service;
+
+        // validator 0 to 7 send Notarize on slot 2
+        let block_id = Hash::new_unique();
+        let notarize_vote = Vote::new_notarization_vote(2, block_id);
+        for my_rank in 0..8 {
+            let vote_keypair = &setup_result.validator_keypairs[my_rank].vote_keypair;
+            let bls_keypair =
+                BLSKeypair::derive_from_signer(vote_keypair, BLS_KEYPAIR_DERIVE_SEED).unwrap();
+            let vote_serialized = bincode::serialize(&notarize_vote).unwrap();
+            let consensus_message = ConsensusMessage::Vote(VoteMessage {
+                vote: notarize_vote,
+                signature: bls_keypair.sign(&vote_serialized).into(),
+                rank: my_rank as u16,
+            });
+            consensus_message_sender.send(consensus_message).unwrap();
+        }
+
+        // Listen on the bls_receiver for certificates
+        wait_for_event(&setup_result.bls_receiver, |event| {
+            if let BLSOp::PushCertificate { certificate } = event {
+                assert_eq!(certificate.certificate.slot(), 2);
```

Comment:
can reuse that slot variable here and everywhere else we have an assert

Reply:
Done.